src/control/system/raft/database.go

```diff
@@ -898,14 +907,41 @@ func (db *Database) FindPoolServiceByLabel(label string) (*system.PoolService, e
 	return nil, system.ErrPoolLabelNotFound(label)
 }
 
+func (db *Database) TakePoolLock(ctx context.Context, uuid uuid.UUID) (*PoolLock, error) {
+	if err := db.CheckLeader(); err != nil {
+		return nil, err
+	}
+	db.Lock()
+	defer db.Unlock()
+
+	lock, err := getCtxLock(ctx)
+	if err != nil && err != errNoCtxLock {
+		return nil, err
+	} else if lock != nil {
```

Comment:
This smells a bit, it feels more comfortable sticking to the premise that if err exists then we shouldn't be using the value returned from the call (lock). Under what circumstances (error value) will we need to be interrogating the returned lock value and is there a more sane way to do this that doesn't require using the value returned in an error condition?

Reply:
If you look at the implementation of `getCtxLock()`, you'll see that it never returns a non-nil lock and a non-nil error at the same time. I'll admit that the above construction does not make this clear, however. 

The logic is that we check the context for an existing lock. If a lock exists, err will be nil, and we check the lock to see if it's a valid lock for the pool UUID. If no lock exists in the context, then we return the sentinel `errNoCtxLock` error with a nil lock and proceed down to taking a new lock on the pool.

I'll see if I can clean this up a bit to reduce the ambiguity.